src/hotspot/share/gc/shenandoah/shenandoahReferenceProcessor.hpp

```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright (c) 2015, 2019, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2020, Red Hat, Inc. and/or its affiliates.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ *
+ */
+
+#ifndef SHARE_VM_GC_SHENANDOAH_SHENANDOAHREFERENCEPROCESSOR_HPP
+#define SHARE_VM_GC_SHENANDOAH_SHENANDOAHREFERENCEPROCESSOR_HPP
+
+#include "gc/shared/referenceDiscoverer.hpp"
+#include "memory/allocation.hpp"
+
+class WorkGang;
+
+static const size_t reference_type_count = REF_PHANTOM + 1;
+typedef size_t Counters[reference_type_count];
+
+/*
+ * Shenandoah concurrent reference processing
+ *
+ * Concurrent reference processing is made up of two main phases:
+ * 1. Concurrent reference marking: Discover all j.l.r.Reference objects and determine reachability of all live objects.
+ * 2. Concurrent reference processing: For all discoved j.l.r.References, determine whether or not to keep or clean
+ *    them. Also, clean and enqueue relevant references concurrently.
+ *
+ * Concurrent reference marking:
+ * The goal here is to establish the kind of reachability for all objects on the heap. We distinguish two kinds of
+ * reachability:
+ * - An object is 'strongly reachable' if it can be found by searching transitively from GC roots.
+ * - An object is 'finalizably reachable' if it is not strongly reachable, but can be found by searching
+ *   from the referents of FinalReferences.
+ *
+ * These reachabilities are implemented in shenandoahMarkBitMap.*
+ * Conceptually, marking starts with a strong wavefront at the GC roots. Whenever a Reference object is encountered,
+ * that Reference is discovered, it may be discovered by the ShenandoahReferenceProcessor. If it is discovered, it
```

Comment:
"Whenever a Reference object is encountered, it may be discovered by the ShenandoahReferenceProcessor"?